### ISSUE TITLE
LYWSD03MMC (Xiaomi firmware) - try fetching precise battery value

### ISSUE THREAD

**Issue by denilsonsa** [custom-components/ble_monitor#478]
## Short Summary

The [Telink Flasher][flasher] can display the precise battery voltage of LYWSD03MMC devices even on the stock firmware. However, the Home Assistant component shows either 100% or 10%.

Please investigate if it is possible to get the precise battery readings in the Home Assistant component.

[flasher]: https://pvvx.github.io/ATC_MiThermometer/TelinkMiFlasher.html

## Context and long description

I have a handful of [LYWSD03MMC](https://github.com/custom-components/ble_monitor/blob/master/pictures/LYWSD03MMC.jpg) sensors, some of them with their original Xiaomi firmware, others with custom [pvvx/ATC firmware](https://github.com/pvvx/ATC_MiThermometer). I know the custom firmware can report very precise battery percentages (or battery voltage), which is great. However, the stock Xiaomi firmware seems to only report 100% or 10%.

![Lovelace screenshot showing battery percentages and voltages](https://user-images.githubusercontent.com/121676/129479208-23816a3a-d508-4201-b3d5-bee98ea62aa4.png)

To my surprise, when I decided to replace the stock firmware of those two devices, I noticed the precise battery voltage in millivolts was shown. This happened before changing the firmware, thus I can conclude the stock firmware can, in fact, report the precise voltage.

![Telink Flasher screenshots, showing the precise battery voltage](https://user-images.githubusercontent.com/121676/129479285-508330c8-62f6-455e-b42b-a21e88ab572f.png)

Can you please investigate if the precise battery readings can be made available in the Home Assistant component?

I don't know if that precise voltage is only available when the device is actively "connected" through Bluetooth, or if it is also available on the "low energy messages" that are broadcast every few minutes. If that's really the case, then I understand this cannot be fixed, and this issue can be closed. I'm sure you have a much better understanding of these devices and their protocol, and that's why I'm bringing it to your attention.

---

Sidenote 1: I'm using the version 1.9.2. Yes, it's a couple of months old, I'll update at a later date, when I have more time.

Sidenote 2: The custom firmware has the following features: ① reporting data more often (about every minute instead of every ten minutes); ② more precise data (extra decimal digits); ③ display the clock (or the battery percentage) on the LCD display; ④ precise battery readings (instead of just "full" and "low"). Of those four features, ①②③ are nice to have, but I can live without them; while ④ is very useful to monitor the remaining charge of the batteries.

**Comment by Ernst79** [comment]
The stock firmware sends only battery (in %) in the passive BLE advertisements (0x0A object). Only after connecting it will send additional info, like voltage level unfortunately. Problem is that with a connection, data is only send to the device that is connected, and can’t be read by other devices/apps at the same time. ATC firmware solves this,as they added the voltage in the passive advertisements, giving you the exact number.

For LYWSD03MMC we had to do a trick, as this sensor could send data with (ATC) or without (Xiaomi) voltage, so when using xiaomi firmware, we convert % to voltage, to prevent an unknown value in HA, but this is still very inaccurate (as you have found out with either 10% or 100%). I have never found the logic how xiaomi determines the % in their messages. 

I would recommend to use the ATC firmware, as it has many advantages as you explained yourself and isn’t difficult to upload with their webtool.

**Comment by denilsonsa** [comment]
Okay, that explains it. I'm gonna close this issue. Thanks!
